### PR TITLE
Refactor symlink code and add commentary and a better human error.

### PR DESCRIPTION
**Pull request by nathanhammond**
This is significantly more verbose but doing algebraic reduction on this in the past has already resulted in at least one bug.

I've added tons of commentary to make it clear for future travelers (read: me in six months).

### CODE REVIEW

**Comment by stefanpenner** [comment]
This looks like it will regress performance, please be mindful of the original implementations goal of mitigation of unneeded disk IO.

File: index.js [hunk 1]
```diff
@@ -177,27 +177,55 @@ Funnel.prototype.build = function() {
...
+     * 3. We need slightly different behavior on rebuilds.
+     *
+     * Behavior has been modified to always having an `else` clause so that
+     * the code is forced to account for all scenarios. Not accounting for
+     * all scenarios made it possible for initial builds to succeed without
+     * specifying `this.allowEmpty`.
+     */
+
+    var inputPathExists = fs.existsSync(inputPath);
+    var outputPathExists = fs.existsSync(this.outputPath);
```
Comment (nathanhammond) [inline index.js]: This cannot use the `exists-sync` npm package in that it is _specifically_ looking for broken symlinks.

Comment (stefanpenner) [inline index.js]: maybe we should use `statSync` `lStatSync`

Comment (nathanhammond) [inline index.js]: Hrm. Tests still pass using `exists-sync`. Not sure if that's a gap in our coverage or... ?

**Comment by nathanhammond** [comment]
Since I managed to get a reproduction without a good test case yet, here is a test that has different behavior in 1.0.6 and 1.0.7: (`git clone && npminstall && npm test`)

- [No error thrown in 1.0.6](https://github.com/broccolijs/broccoli-funnel/compare/v1.0.6...nathanhammond:failure-in-1.0.6?expand=1)
- [Error thrown in 1.0.7](https://github.com/broccolijs/broccoli-funnel/compare/v1.0.7...nathanhammond:failure-in-1.0.7?expand=1)

```
     The Broccoli Plugin: [Funnel] failed with:
  Error: ENOENT: no such file or directory, lstat '/Users/nathanhammond/Repos/broccoli-funnel/tmp/funnel-output_path-CYpaNO7K.tmp'
      at Error (native)
      at Object.fs.lstatSync (fs.js:839:18)
      at symlink (node_modules/symlink-or-copy/index.js:60:26)
      at symlinkOrCopySync (node_modules/symlink-or-copy/index.js:55:5)
      at node_modules/broccoli-plugin/read_compat.js:58:9
      at tryCatch (node_modules/rsvp/dist/rsvp.js:538:12)
      at invokeCallback (node_modules/rsvp/dist/rsvp.js:553:13)
      at publish (node_modules/rsvp/dist/rsvp.js:521:7)
      at flush (node_modules/rsvp/dist/rsvp.js:2373:5)
  
  The broccoli plugin was instantiated at: 
      at Funnel.Plugin (node_modules/broccoli-plugin/index.js:7:31)
      at new Funnel (index.js:56:10)
      at Context.<anonymous> (tests/index.js:339:15)
```

The error is "correct", but:
- Only shows up on rebuild.
- Can be detected and have a better human-readable error message.

**Comment by nathanhammond** [comment]
> This looks like it will regress performance, please be mindful of the original implementations goal of mitigation of unneeded disk IO.

The only difference in terms of IO effort I can identify is no short-circuiting on `(inputPathExist && !fs.existsSync(this.outputPath))` which instead gets invoked each time. I can add back in a guard for that.

The remaining code has no additional invocations of anything as a consequence of this change. The increased verbosity is simply to make the code tractable. It is now _much_ easier to confirm via manual review that we do the minimum amount of work necessary for each of the six scenarios–the situation for each is described inline.

**Comment by nathanhammond** [comment]
Test case up, ready for review/merge.

**Comment by stefanpenner** [comment]
ah the bug was actually broken input, and we didn't detect an error fast. Sounds good!

**Pull request merged**